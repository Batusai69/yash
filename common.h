/* Yash: yet another shell */
/* common.h: defines symbols common to all sources. */
/* (C) 2007-2010 magicant */

/* This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 2 of the License, or
 * (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.  */

/* This file should be included at the very first in every source file. */


#ifndef YASH_COMMON_H
#define YASH_COMMON_H

#if HAVE_CONFIG_H
# include "config.h"
#endif

#ifndef _POSIX_C_SOURCE
# define _POSIX_C_SOURCE 200112L
#endif
#ifndef _XOPEN_SOURCE
# define _XOPEN_SOURCE 600
#endif

#ifndef __GNUC__
# define __attribute__(ignore)
#endif

#ifdef __CYGWIN__
# undef __STRICT_ANSI__
#endif

#if HAVE_GETTEXT
# define gt(MSGID)  ((const char *) gettext(MSGID))
# define Ngt(MSGID) MSGID
# if HAVE_NGETTEXT
#  define ngt(MSGID,MSGS,MSGP,N) ((const char *) ngettext(MSGS, MSGP, N))
# else
#  define ngt(MSGID,MSGS,MSGP,N) gt(MSGID)
# endif
#else
# define gt(MSGID)              MSGID
# define Ngt(MSGID)             MSGID
# define ngt(MSGID,MSGS,MSGP,N) MSGID
#endif

#define ARGV(i) ((wchar_t *) argv[i])

#endif /* YASH_COMMON_H */


/* vim: set ts=8 sts=4 sw=4 noet tw=80: */
