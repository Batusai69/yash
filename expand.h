/* Yash: yet another shell */
/* expand.h: functions for command line expansion */
/* Â© 2007-2008 magicant */

/* This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 2 of the License, or
 * (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.  */


#ifndef EXPAND_H
#define EXPAND_H

#include <stdbool.h>
#include "util.h"


bool expand_line(char *const *args, int *argc, char ***argv)
	__attribute__((nonnull(2,3)));
char *expand_single(const char *arg)
	__attribute__((nonnull));
char *expand_word(const char *s)
	__attribute__((nonnull));
void add_splitting(const char *str, struct strbuf *buf, struct plist *list,
		const char *ifs, const char *q)
	__attribute__((nonnull(1,2,3,4)));
void escape_sq(const char *s, struct strbuf *buf)
	__attribute__((nonnull));
void escape_bs(const char *s, const char *q, struct strbuf *buf)
	__attribute__((nonnull));


#endif /* EXPAND_H */
