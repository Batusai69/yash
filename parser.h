/* Yash: yet another shell */
/* parser.h: interface to the parser */
/* © 2007-2008 magicant */

/* This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 2 of the License, or
 * (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.  */


#ifndef PARSER_H
#define PARSER_H

#include <stdbool.h>
#include "lineinput.h"


/* 解析中のソースコードに関する情報を表す */
struct parse_info {
	const char *filename;
	unsigned lineno;
	getline_t *input;
	void *inputinfo;
};
/* filename はソースコードのファイル名。エラー表示で使う。NULL でもよい。
 * lineno は行番号。解析が進むにつれて増える。最初は 0。
 * input は入力関数。 */

typedef struct x_redirect  REDIR;
typedef struct x_process   PROCESS;
typedef struct x_pipeline  PIPELINE;
typedef struct x_statement STATEMENT;

/* リダイレクトを表す */
struct x_redirect {
	REDIR *next;
	enum redirect_type {
		RT_INPUT, RT_OUTPUT, RT_OUTCLOB, RT_APPEND, RT_INOUT,
		RT_DUPIN, RT_DUPOUT, RT_HERE,    RT_HERERT,
	}      rd_type;
	int    rd_fd;           /* リダイレクトするファイルディスクリプタ */
	char  *rd_value;        /* リダイレクトの引数 (常に非 NULL) */
	char  *rd_herecontent;  /* ヒアドキュメントの内容 */
};
/* 例えば '2>&1' では、rd_fd = 2, rd_value = "1" となる。 */
/* rd_herecontent は実際のヒアドキュメントの内容。RT_HERERT では、内容を
 * 読み取った段階で予め行頭のタブを除いておく。 */

/* パイプラインに含まれる一つのプロセス */
struct x_process {
	PROCESS       *next;       /* パイプライン内の次のプロセス */
	enum {
		PT_NORMAL,    /* 普通のコマンド */
		PT_GROUP,     /* 現在のシェルで実行するコマンド群: { ... } */
		PT_SUBSHELL,  /* サブシェルで実行するコマンド群: ( ... ) */
	}              p_type;
	char         **p_assigns;  /* 変数代入 */
	union {
		char     **p_args;     /* コマンド名と引数 */
		STATEMENT *p_subcmds;  /* プロセスに含まれる文の内容 */
	} p_content_;
	REDIR         *p_redirs;   /* プロセスに対するリダイレクト */
};
#define p_args    p_content_.p_args
#define p_subcmds p_content_.p_subcmds
/* p_type が PT_NORMAL のとき、p_args にコマンドの内容 (変数代入やリダイレクトを
 * 除く) が入る。(空白ごとに分けてエイリアスを展開しただけで、
 * それ以上パラメータの展開などは行っていない)
 * p_type が非 PT_NORMAL のときは、プロセスに含まれるサブステートメントが
 * p_subcmds に入る。
 * p_redirs はリダイレクトのデータを表す。(こちらも展開していない) */

/* 一つのパイプライン */
struct x_pipeline {
	PIPELINE *next;     /* コマンドリスト内の次のパイプライン */
	PROCESS  *pl_proc;  /* パイプラインの最初のプロセス */
	bool      pl_neg;   /* パイプラインの終了ステータスを反転するかどうか */
	bool      pl_loop;  /* パイプラインが環状かどうか */
	bool      pl_next_cond;
};
/* pl_next_cond は next を実行する条件。
 * true は '&&' を、false は '||' を意味する */

/* 一つの文 */
struct x_statement {
	STATEMENT *next;        /* ソース内の次の文 */
	PIPELINE  *s_pipeline;  /* 文の中の最初のパイプライン */
	bool       s_bg;        /* バックグラウンドかどうか */
};


extern unsigned lineno;

int read_and_parse(
		struct parse_info *restrict info, STATEMENT **restrict result)
	__attribute__((nonnull));
char *skip_with_quote(const char *restrict s, const char *restrict delim)
	__attribute__((nonnull));
char *skip_without_quote(const char *restrict s, const char *restrict delim)
	__attribute__((nonnull));
char *make_statement_name(PIPELINE *p);
char *make_pipeline_name(PIPELINE *p);
void procsfree(PROCESS *processes);
void pipesfree(PIPELINE *pipelines);
void statementsfree(STATEMENT *statements);


#endif /* PARSER_H */
