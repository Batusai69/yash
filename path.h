/* Yash: yet another shell */
/* path.h: filename-related utilities */
/* © 2007-2008 magicant */

/* This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 2 of the License, or
 * (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.  */


#ifndef PATH_H
#define PATH_H

#include <stdbool.h>
#include <stddef.h>
#include <dirent.h>


extern bool is_readable(const char *path)
    __attribute__((nonnull));
extern bool is_executable(const char *path)
    __attribute__((nonnull));
extern bool is_directory(const char *path)
    __attribute__((nonnull));

extern bool is_same_file(const char *path1, const char *path2)
    __attribute__((nonnull));

extern wchar_t *canonicalize_path(const wchar_t *path)
    __attribute__((nonnull,malloc,warn_unused_result));

extern char *which(
	const char *restrict name,
	char *const *restrict dirs,
	bool cond(const char *path))
    __attribute__((nonnull(1),malloc,warn_unused_result));

extern int xclosedir(DIR *dir)
    __attribute__((nonnull));


/********** patharray **********/

extern void reset_patharray(const char *newpath);


/********** コマンド名ハッシュ **********/

extern void init_cmdhash(void);
extern const char *get_command_path(const char *name, bool forcelookup)
    __attribute__((nonnull));
extern void fill_cmdhash(const char *prefix, bool ignorecase);


/********** ホームディレクトリキャッシュ **********/

extern struct passwd *xgetpwnam(const char *name)
    __attribute__((nonnull));
extern void init_homedirhash(void);
extern void clear_homedirhash(void);
extern const wchar_t *get_home_directory(
	const wchar_t *username, bool forcelookup)
    __attribute__((nonnull));


/********** wglob **********/

enum wglbflags {
    WGLB_MARK     = 1 << 0,
    WGLB_NOESCAPE = 1 << 1,
    WGLB_CASEFOLD = 1 << 2,
    WGLB_PERIOD   = 1 << 3,
    WGLB_NOSORT   = 1 << 4,
    WGLB_RECDIR   = 1 << 5,
};

struct plist_T;

extern bool wglob(const wchar_t *restrict pattern, enum wglbflags flags,
	struct plist_T *restrict list)
    __attribute__((nonnull));


#endif /* PATH_H */


/* vim: set ts=8 sts=4 sw=4 noet: */
