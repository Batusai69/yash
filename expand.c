/* Yash: yet another shell */
/* expand.c: word expansion */
/* © 2007-2008 magicant */

/* This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 2 of the License, or
 * (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.  */


#include "common.h"
#include <assert.h>
#include <stdbool.h>
#include <stdlib.h>
#include "util.h"
#include "strbuf.h"
#include "plist.h"
#include "path.h"
#include "parser.h"
#include "variable.h"
#include "expand.h"


#define ESCAPED_CHARS L"*?[{"  /* ブレース展開・glob で特殊な意味を持つ文字 */

__attribute__((nonnull(4)))
static bool expand_word(const wordunit_T *arg,
	tildetype_T tilde, const wchar_t *ifs, plist_T *list);
__attribute__((nonnull,malloc,warn_unused_result))
static wchar_t *expand_tilde(const wchar_t **ss);
__attribute__((nonnull,malloc,warn_unused_result))
static void **expand_param(const paramexp_T *p, bool indq, const wchar_t *ifs);
__attribute__((nonnull))
static wchar_t *command_substitution(const wchar_t *code);
__attribute__((nonnull))
static inline void add_sq(const wchar_t *restrict *ss, xwcsbuf_T *restrict buf);
__attribute__((nonnull(2,3)))
static void add_splitting(wchar_t *s,
	plist_T *list, xwcsbuf_T *buf, const wchar_t *ifs);


/* コマンドライン上の各種展開を行う。
 * args: void * にキャストした const wordunit_T へのポインタの配列。配列内の
 *       各ワードが展開される。配列の最後の要素は NULL でなければならない。
 * argcp: このポインタが指すところに展開結果の個数が入る。
 * argvp: このポインタが指すところに展開結果が入る。結果は、マルチバイト文字列
 *       へのポインタの配列へのポインタであり、配列の最後の要素は NULL である。
 * args で与えられる内容は変更されない。
 * 戻り値: 成功すると true、エラーがあると false。
 * エラーがあった場合、*argcp, *argvp の値は不定である */
bool expand_line(void *const *restrict args,
    int *restrict argcp, char ***restrict argvp)
{
    plist_T list1, list2;
    const wchar_t *ifs = getvar(VAR_IFS);

    pl_init(&list1);

    /* 各種展開をする */
    while (*args) {
	if (!expand_word(*args, tt_single, ifs, &list1)) {
	    recfree(pl_toary(&list1), free);
	    return false;
	}
	args++;
    }

    /* ブレース展開する */
    if (false) {  // TODO expand: expand_line: ブレース展開
    } else {
	list2 = list1;
    }

    /* glob する */
    if (false) {  // TODO expand: expand_line: no_glob オプション
	for (size_t i = 0; i < list1.length; i++)
	    list1.contents[i] = realloc_wcstombs(list2.contents[i]);
	list1 = list2;
    } else {
	enum wglbflags flags = 0;
	pl_init(&list1);
	// TODO expand: expand_line: wglob のフラグオプション
	for (size_t i = 0; i < list2.length; i++) {
	    size_t oldlen = list1.length;
	    wglob(list2.contents[i], flags, &list1);
	    if (oldlen == list1.length)
		pl_add(&list1, realloc_wcstombs(unescape(list2.contents[i])));
	    free(list2.contents[i]);
	}
	pl_destroy(&list2);
    }

    *argcp = list1.length;
    *argvp = (char **) pl_toary(&list1);
    return true;
}

/* 一つの単語を展開する。
 * 各種展開と引用符除去を行う。フィールド分割はしない。
 * glob が true ならファイル名展開も行うが、結果が一つでない場合は……
 *   - posixly_correct が true ならファイル名展開前のパターンを返し、
 *   - posixly_correct が false ならエラーとする。
 * エラー発生時はメッセージを出して NULL を返す。 */
wchar_t *expand_single(const wordunit_T *arg, tildetype_T tilde, bool glob);

/* 一つの文字列を展開する。
 * パラメータ展開・数式展開・コマンド置換を行うが、引用符除去・フィールド分割・
 * ファイル名展開はしない。エラー発生時はメッセージを出して NULL を返す。 */
wchar_t *expand_string(const wordunit_T *arg);

/* 一つの単語を展開する。
 * チルダ展開・パラメータ展開・コマンド置換・数式展開をし、単語分割をする。
 * w:      展開する単語
 * tilde:  チルダ展開の種類
 * ifs:    非 NULL なら、これに従って単語分割する。
 * list:   結果を入れるリスト
 * 戻り値: エラーがなければ true。
 * ifs が NULL なら、結果は常に 1 要素である。
 * 引用符 (" と ') はバックスラッシュエスケープに置き換わる。 */
bool expand_word(const wordunit_T *w,
	tildetype_T tilde, const wchar_t *ifs, plist_T *list)
{
    bool ok = true;
    bool indq = false;     /* 二重引用符 " の中かどうか */
    bool first = true;     /* 最初の word unit かどうか */
    bool force = false;    /* 展開結果が空文字列でも追加する */
    bool suppress = false; /* force を無効にする */
    size_t initlen = list->length;
    xwcsbuf_T buf;
    const wchar_t *str;
    wchar_t *s;
    void **array;

    wb_init(&buf);
    while (w) {
	switch (w->wu_type) {
	case WT_STRING:
	    str = w->wu_string;
	    if (first && tilde != tt_none)
		add_splitting(expand_tilde(&str), list, &buf, ifs);
	    for (;;) {
		switch (*str) {
		case L'\0':
		    goto out;
		case L'"':
		    indq = !indq;
		    force = true;
		    break;
		case L'\'':
		    if (indq)
			goto default_case;
		    force = true;
		    add_sq(&str, &buf);
		    break;
		case L'\\':
		    wb_wccat(&buf, L'\\');
		    if (*++str)
			wb_wccat(&buf, *str++);
		    continue;
		case L':':
		    if (!indq && tilde == tt_multi) {
			wb_wccat(&buf, L':');
			str++;
			add_splitting(expand_tilde(&str), list, &buf, ifs);
			continue;
		    }
		    /* falls thru! */
		default:  default_case:
		    if (indq)
			wb_wccat(&buf, L'\\');
		    wb_wccat(&buf, *str);
		    break;
		}
		str++;
	    }
out:
	    break;
	case WT_PARAM:
	    array = expand_param(w->wu_param, indq, ifs);
	    if (array) {
		if (!*array) {
		    suppress = true;
		} else {
		    const wchar_t *iifs = indq ? L"" : ifs;
		    for (void **a = array; ; ) {
			add_splitting(*a, list, &buf, iifs);
			if (!*++a)
			    break;
			if (!iifs || buf.length > 0) {
			    pl_add(list, wb_towcs(&buf));
			    wb_init(&buf);
			}
		    }
		}
		free(array);
	    } else {
		ok = false;
	    }
	    break;
	case WT_CMDSUB:
	    s = command_substitution(w->wu_cmdsub);
	    if (s) {
		add_splitting(escape(s, ESCAPED_CHARS), list, &buf,
			indq ? L"" : ifs);
		free(s);
	    } else {
		ok = false;
	    }
	    break;
	case WT_ARITH:
	    ok = false;  // TODO expand: expand_word: 数式展開の実装
	    xerror(0, 0, "arithmetic expansion not implemented");
	    break;
	}
	w = w->next;
	first = false;
    }

    if (!ifs || buf.length > 0
	    || (initlen == list->length && force && !suppress))
	pl_add(list, wb_towcs(&buf));
    else
	wb_destroy(&buf);

    return ok;
}

/* 単一引用符の中身を、エスケープしながら加える。
 * 最初 *ss は開く引用符を指していて、返るとき *ss は閉じる引用符を指す。 */
void add_sq(const wchar_t *restrict *ss, xwcsbuf_T *restrict buf)
{
    (*ss)++;
    for (;;) {
	switch (**ss) {
	    case L'\0':
		assert(false);
	    case L'\'':
		return;
	    case L'*':  case L'?':  case L'[':  case L'{':  /* ESCAPED_CHARS */
		wb_wccat(buf, L'\\');
		/* falls thru! */
	    default:
		wb_wccat(buf, **ss);
		break;
	}
	(*ss)++;
    }
}

/* チルダ展開を行う。
 * ss: チルダがあるべき場所を指すポインタへのポインタ。
 *     *ss は展開した後の部分まで進む。
 * 戻り値: 新しく malloc した、展開結果。失敗なら NULL。
 * (*ss)[0] が L'~' でなければ直ちに NULL を返す。 */
wchar_t *expand_tilde(const wchar_t **ss)
{
    // TODO expand:expand_tilde
    (void) ss;
    return NULL;
}

/* パラメータ展開を行い、結果を返す。
 * 戻り値: 展開結果。void * にキャストしたワイド文字列へのポインタの NULL 終端
 *         配列。配列および要素は新しく malloc したものである。
 *         エラーのときは NULL。
 * 返す各要素は、ESCAPED_CHARS をエスケープ済みである。
 * "@" または配列以外の展開結果は、必ず要素数 1 である。
 * "*" の展開結果は、ifs に従って結合済みである。 */
void **expand_param(const paramexp_T *p, bool indq, const wchar_t *ifs)
{
    (void) p, (void) indq, (void) ifs;
    // TODO expand:expand_param
    return NULL;
}

/* コマンド置換を行い、結果を返す。
 * 戻り値: 新しく malloc した、置換結果の文字列。エラーなら NULL。 */
wchar_t *command_substitution(const wchar_t *code)
{
    (void) code;
    // TODO expand:command_substitution
    return NULL;
}

/* 単語をバッファに追加する。
 * s:    追加する単語。この関数内で free する。NULL なら何もしない。
 * list: 単語分割した場合に結果を追加するリスト
 * buf:  s を追加するバッファ
 * ifs:  非 NULL なら s を追加する際にこれに従って単語分割する。 */
void add_splitting(wchar_t *s,
	plist_T *list, xwcsbuf_T *buf, const wchar_t *ifs)
{
    free(s);
    // TODO expand:add_splitting
    (void) list, (void) buf, (void) ifs;
}

/* s の文字のうち、t に含まれるものをバックスラッシュエスケープして、
 * 新しく malloc した文字列として返す。
 * 戻り値: エスケープした s */
wchar_t *escape(const wchar_t *restrict s, const wchar_t *restrict t)
{
    xwcsbuf_T buf;
    wb_init(&buf);
    while (*s) {
	if (wcschr(t, *s))
	    wb_wccat(&buf, L'\\');
	wb_wccat(&buf, *s);
	s++;
    }
    return wb_towcs(&buf);
}

/* バックスラッシュエスケープを削除して新しく malloc した文字列として返す。 */
wchar_t *unescape(const wchar_t *s)
{
    xwcsbuf_T buf;
    wb_init(&buf);
    while (*s) {
	if (*s == L'\\') {
	    if (*(s+1) == L'\0')
		break;
	    else
		s++;
	}
	wb_wccat(&buf, *s);
	s++;
    }
    return wb_towcs(&buf);
}


/* vim: set ts=8 sts=4 sw=4 noet: */
