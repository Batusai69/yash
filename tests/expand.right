expand.test
1 b
123/xxx/789
xxx/456/789 123/456/789
123/456/xxx 123/456/789
123-456-789 123/4/6/789 123/456/789
x
19
1 2 3 4 5
