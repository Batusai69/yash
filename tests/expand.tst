[ ~+ = $PWD ] && echo \~+
# TODO needs 'cd' builtin
# cd .
# [ ~- = $OLDPWD ] && echo \~-

var=123/456/789 asterisks='*****'

echo expand.t${{asterisks#"**"}%"**"}st
echo ${#$(echo a)} ${${`echo abc`#a}%c}

echo ${var/456/xxx}
echo ${var/#123/xxx} ${var/#456/xxx}
echo ${var/%789/xxx} ${var/%456/xxx}
echo ${var//\//-} ${var//5//} ${var///-}
echo ${var:/123*789/x}

$TESTEE -c 'echo ${#*}' x 1 22 333 4444 55555
$TESTEE -c 'echo ${#@}' x 1 22 333 4444 55555

echo $((echo foo); (echo bar))

# TODO needs 'unset'
# unset IFS
# var=1-2-3
# echo $var ${IFS= -} $var
