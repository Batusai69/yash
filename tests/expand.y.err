===== 11 =====
../yash: arithmetic: division by zero
../yash: arithmetic: division by zero
../yash: arithmetic: division by zero
../yash: arithmetic: division by zero
../yash: arithmetic: division by zero
../yash: arithmetic: division by zero
../yash: arithmetic: division by zero
../yash: arithmetic: division by zero
