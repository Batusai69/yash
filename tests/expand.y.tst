[ ~+ = $PWD ] && echo \~+
cd .
[ ~- = $OLDPWD ] && echo \~-

var=123/456/789 asterisks='*****'

echo expand.y.t${{asterisks#"**"}%"**"}st
echo ${#$(echo a)} ${${`echo abc`#a}%c}
[ x"${$+x}" != x"${$:+y}" ] && echo nest \$ +
[ x"${$-x}" =  x"${$:-y}" ] && echo nest \$ -
[ x"${$=x}" =  x"${$:=y}" ] && echo nest \$ =
[ x"${$?x}" =  x"${$:?y}" ] && echo nest \$ ?
[ x"${$#x}" =  x"${$##y}" ] && echo nest \$ \#
[ x"${$%x}" =  x"${$%%y}" ] && echo nest \$ %
[ x"${$/x/y}" = x"${$:/x/y}" ] && echo nest \$ /

echo ${var/456/xxx}
echo ${var/#123/xxx} ${var/#456/xxx}
echo ${var/%789/xxx} ${var/%456/xxx}
echo ${var//\//-} ${var//5//} ${var///-}
echo ${var:/123*789/x}

$INVOKE $TESTEE -c 'echo "${#*}"' x 1 22 333 4444 55555
$INVOKE $TESTEE -c 'echo "${#@}"' x 1 22 333 4444 55555

echo $((echo foo); (echo bar))

var=1-2-3 IFS=
echo $var ${IFS:= -} $var
unset IFS

echo ===== 1 =====

echo "${var[2,4]}"
echo "${var[3,3]}" "${var[3]}"
echo "${var[4,3]}"
(echo "${var[0]}") 2>/dev/null || echo zero index
echo "${var[@]}" "${var[*]}" ${var[#]}

echo ===== 2 =====

echol () {
	typeset i  # local
	for i; do printf "%s\n" "$i"; done
}

ary=(1 22 '3  3' 4"   "4 5)
echol "${ary}" -0-
echol "${ary[@]}" -1-
echol "${ary[*]}" -2-
echol "${ary[#]}" -3-
echol "${ary[1]}" -4-
echol "${ary[2]}" -5-
echol "${ary[3]}" -6-
echol "${ary[4]}" -7-
echol "${ary[5]}" -8-
echol "${ary[1,3]}" -9-
echol "${ary[2,4]}" -10-
echol "${ary[3,5]}" -11-
echol "${ary[4,6]}" -12-
echol "${ary[5,7]}" -13-
echol "${ary[6,8]}" -14-
echol "${${ary}}" -20-
echol "${${ary[3]}[1]}" -21-
echol "${${ary[2,4]}[2]}" -22-
echol ${ary} -30-
echol ${ary[@]} -31-

echo ===== 3 =====

set "$ary"
unset -v ary
echo "-$ary-"
echol "${@}" -0-
echol "${@[@]}" -1-
echol "${@[*]}" -2-
echol "${@[#]}" -3-
echol "${@[1]}" -4-
echol "${@[2]}" -5-
echol "${@[3]}" -6-
echol "${@[4]}" -7-
echol "${@[5]}" -8-
echol "${@[1,3]}" -9-
echol "${@[2,4]}" -10-
echol "${@[3,5]}" -11-
echol "${@[4,6]}" -12-
echol "${@[5,7]}" -13-
echol "${@[6,8]}" -14-
echol "${${@}}" -20-
echol "${${@[3]}[1]}" -21-
echol "${${@[2,4]}[2]}" -22-
echol ${@} -30-
echol ${@[@]} -31-

echo ===== 4 =====

echol "${*}" -0-
echol "${*[@]}" -1-
echol "${*[*]}" -2-
echol "${*[#]}" -3-
echol "${*[1]}" -4-
echol "${*[2]}" -5-
echol "${*[3]}" -6-
echol "${*[4]}" -7-
echol "${*[5]}" -8-
echol "${*[1,3]}" -9-
echol "${*[2,4]}" -10-
echol "${*[3,5]}" -11-
echol "${*[4,6]}" -12-
echol "${*[5,7]}" -13-
echol "${*[6,8]}" -14-
echol "${${*}}" -20-
echol "${${*[3]}[1]}" -21-
echol "${${*[2,4]}[2]}" -22-
echol ${*} -30-
echol ${*[@]} -31-

echo ===== 5 =====

echol "${@[1 + (4 / 2), 5 - 1]}"

echo ===== 6 =====

ary=('' '' '')
printf '[%s]' "${ary[@]}"; echo
echo - ${ary[1]=1} ${ary[2]=2} ${ary[3]=3}
echo - ${ary[1]:=1}
printf '[%s]' "${ary[@]}"; echo
echo - ${ary[2]:=2}
echo - ${ary[1]:=X}
printf '[%s]' "${ary[@]}"; echo

readonly ary
( echo - ${ary[3]:=3} ) 2>/dev/null
[ $? -ne 0 ]; echo $?
printf '[%s]' "${ary[@]}"; echo

export ary
env | grep ^ary=

echo ===== 7 =====

set -o braceexpand
echo {1..9 {1.9} 1..9}
echo {1..9}
echo {9..11}
echo {111..123}
echo {9..1}
echo {11..9}
echo {123..111}
echo {1..5..2}
echo {1..19..3}
echo {1..2..3}
echo {1..1..1}
echo {0..0..0}
echo {5..1..-2}
echo {19..1..-3}
echo {1..1..-1}
echo {-11..-19..-2}
echo a{1..4}b
echo a{1,2,3,4,,555}b
echo a{1,2}b{3,4}c
echo a{b{1,2}c,3}d
echo a{b{1..2}c,3}d
echo {1,,2}
echo a{1\'\\\"2,3}b
echo \{1,2,3} {4,5,6\}
echo a{,}b
