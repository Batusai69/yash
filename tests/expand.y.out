~+
~-
expand.y.tst
1 b
nest $ +
nest $ -
nest $ =
nest $ ?
nest $ #
nest $ %
nest $ /
123/xxx/789
xxx/456/789 123/456/789
123/456/xxx 123/456/789
123-456-789 123/4/6/789 123/456/789
x
19
1 2 3 4 5
foo bar
1-2-3   1 2 3
===== 1 =====
-2-
2 2

zero index
1-2-3 1-2-3 5
===== 2 =====
1
22
3  3
4   4
5
-0-
1
22
3  3
4   4
5
-1-
1 22 3  3 4   4 5
-2-
5
-3-
1
-4-
22
-5-
3  3
-6-
4   4
-7-
5
-8-
1
22
3  3
-9-
22
3  3
4   4
-10-
3  3
4   4
5
-11-
4   4
5
-12-
5
-13-
-14-
1
22
3  3
4   4
5
-20-
3
-21-
3  3
-22-
1
22
3
3
4
4
5
-30-
1
22
3
3
4
4
5
-31-
===== 3 =====
--
1
22
3  3
4   4
5
-0-
1
22
3  3
4   4
5
-1-
1 22 3  3 4   4 5
-2-
5
-3-
1
-4-
22
-5-
3  3
-6-
4   4
-7-
5
-8-
1
22
3  3
-9-
22
3  3
4   4
-10-
3  3
4   4
5
-11-
4   4
5
-12-
5
-13-
-14-
1
22
3  3
4   4
5
-20-
3
-21-
3  3
-22-
1
22
3
3
4
4
5
-30-
1
22
3
3
4
4
5
-31-
===== 4 =====
1 22 3  3 4   4 5
-0-
1 22 3  3 4   4 5
-1-
1 22 3  3 4   4 5
-2-
5
-3-
1
-4-
22
-5-
3  3
-6-
4   4
-7-
5
-8-
1 22 3  3
-9-
22 3  3 4   4
-10-
3  3 4   4 5
-11-
4   4 5
-12-
5
-13-

-14-
1 22 3  3 4   4 5
-20-
3
-21-
2
-22-
1
22
3
3
4
4
5
-30-
1
22
3
3
4
4
5
-31-
===== 5 =====
3  3
4   4
===== 6 =====
[][][]
-
- 1
[1][][]
- 2
- 1
[1][2][]
0
[1][2][]
ary=1:2:
===== 7 =====
{1..9 {1.9} 1..9}
1 2 3 4 5 6 7 8 9
9 10 11
111 112 113 114 115 116 117 118 119 120 121 122 123
9 8 7 6 5 4 3 2 1
11 10 9
123 122 121 120 119 118 117 116 115 114 113 112 111
1 3 5
1 4 7 10 13 16 19
1
1
{0..0..0}
5 3 1
19 16 13 10 7 4 1
1
-11 -13 -15 -17 -19
a1b a2b a3b a4b
a1b a2b a3b a4b ab a555b
a1b3c a1b4c a2b3c a2b4c
ab1cd ab2cd a3d
ab1cd ab2cd a3d
1  2
a1'\"2b a3b
{1,2,3} {4,5,6}
ab ab
